localstack/utils/files.py

```diff
@@ -139,15 +139,36 @@ def chown_r(path: str, user: str):
 
 
 def chmod_r(path: str, mode: int):
-    """Recursive chmod"""
+    """
+    Recursive chmod
+    :param path: path to file or directory
+    :param mode: permission mask as octal integer value
+    """
     if not os.path.exists(path):
         return
-    os.chmod(path, mode)
+    idempotent_chmod(path, mode)
     for root, dirnames, filenames in os.walk(path):
         for dirname in dirnames:
-            os.chmod(os.path.join(root, dirname), mode)
+            idempotent_chmod(os.path.join(root, dirname), mode)
         for filename in filenames:
-            os.chmod(os.path.join(root, filename), mode)
+            idempotent_chmod(os.path.join(root, filename), mode)
+
+
+def idempotent_chmod(path: str, mode: int):
+    """
+    Perform idempotent chmod on the given file path (non-recursively). The function attempts to call `os.chmod`, and
+    will catch and only re-raise exceptions (e.g., PermissionError) if the file does not have the given mode already.
+    :param path: path to file
+    :param mode: permission mask as octal integer value
+    """
+    try:
+        os.chmod(path, mode)
+    except Exception:
+        existing_mode = os.stat(path)
+        if mode in [existing_mode.st_mode, existing_mode.st_mode & 0o777]:
```

Comment:
Using the `0o777` bitmask here, to select the file permission mode bits only - as `st_mode` contains the combination of file type + file mode bits (e.g., `0o100765`).

Reply:
Why do we have to check for both equivalencies here? Do we expect mode to be full `st_mode` format?

Reply:
Good catch. One pattern that we may see is the following: if we get the current mode via `os.stat(..)` first, and then modify the mode to add the executable bits - basically the equivalent to `chmod a+x /my/file`:

```
    mode = os.stat("/my/file").st_mode
    mode |= 0o111  # add executable flags
    os.fchmod("/my/file", mode)
```

I don't think we currently have this in codebase, but accepting the full format would make us a bit more forward-compatible.

But your comment helped me look into it in more detail again, actually - I did add a small change now to replace the manual bitmask `& 0o777` with [`stat.S_IMODE(..)`](https://docs.python.org/3/library/stat.html#stat.S_IMODE) (this will also include the fourth bit chmod bit, used to, e.g., set the _sticky bit_).